File: decompiled/sources/com/hjq/widget/view/SlantedTextView.java
package com.hjq.widget.view;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import androidx.core.view.GravityCompat;
import com.hjq.widget.R;

public final class SlantedTextView extends View {
    public static final int ROTATE_ANGLE = 45;
    private final Paint mBackgroundPaint;
    private int mColorBackground;
    private int mGravity;
    private String mText;
    private final Rect mTextBounds;
    private int mTextHeight;
    private final TextPaint mTextPaint;
    private boolean mTriangle;

    public SlantedTextView(Context context) {
        this(context, (AttributeSet) null);
    }

    public SlantedTextView(Context context, AttributeSet attributeSet) {
        this(context, attributeSet, 0);
    }

    public SlantedTextView(Context context, AttributeSet attributeSet, int i) {
        super(context, attributeSet, i);
        this.mText = "";
        this.mTextBounds = new Rect();
        Paint paint = new Paint();
        this.mBackgroundPaint = paint;
        paint.setStyle(Paint.Style.FILL);
        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_OVER));
        paint.setAntiAlias(true);
        TextPaint textPaint = new TextPaint(1);
        this.mTextPaint = textPaint;
        textPaint.setAntiAlias(true);
        TypedArray obtainStyledAttributes = context.obtainStyledAttributes(attributeSet, R.styleable.SlantedTextView);
        setText(obtainStyledAttributes.getString(R.styleable.SlantedTextView_android_text));
        setTextSize(0, (float) obtainStyledAttributes.getDimensionPixelSize(R.styleable.SlantedTextView_android_textSize, (int) getResources().getDimension(R.dimen.sp_12)));
        setTextColor(obtainStyledAttributes.getColor(R.styleable.SlantedTextView_android_textColor, -1));
        setTextStyle(Typeface.defaultFromStyle(obtainStyledAttributes.getInt(R.styleable.SlantedTextView_android_textStyle, 0)));
        setGravity(obtainStyledAttributes.getInt(R.styleable.SlantedTextView_android_gravity, GravityCompat.END));
        setColorBackground(obtainStyledAttributes.getColor(R.styleable.SlantedTextView_android_colorBackground, getAccentColor()));
        setTriangle(obtainStyledAttributes.getBoolean(R.styleable.SlantedTextView_triangle, false));
        obtainStyledAttributes.recycle();
    }

    /* access modifiers changed from: protected */
    public void onMeasure(int i, int i2) {
        int i3;
        TextPaint textPaint = this.mTextPaint;
        String str = this.mText;
        int i4 = 0;
        textPaint.getTextBounds(str, 0, str.length(), this.mTextBounds);
        this.mTextHeight = this.mTextBounds.height() + getPaddingTop() + getPaddingBottom();
        int mode = View.MeasureSpec.getMode(i);
        if (mode == Integer.MIN_VALUE || mode == 0) {
            i3 = this.mTextBounds.width() + getPaddingLeft() + getPaddingRight();
        } else if (mode != 1073741824) {
            i3 = 0;
        } else {
            i3 = View.MeasureSpec.getSize(i);
        }
        int mode2 = View.MeasureSpec.getMode(i2);
        if (mode2 == Integer.MIN_VALUE || mode2 == 0) {
            i4 = this.mTextBounds.height() + getPaddingTop() + getPaddingBottom();
        } else if (mode2 == 1073741824) {
            i4 = View.MeasureSpec.getSize(i2);
        }
        setMeasuredDimension(Math.max(i3, i4), Math.max(i3, i4));
    }

    /* access modifiers changed from: protected */
    public void onDraw(Canvas canvas) {
        drawBackground(canvas);
        drawText(canvas);
    }

    private void drawBackground(Canvas canvas) {
        Path path = new Path();
        int width = canvas.getWidth();
        int height = canvas.getHeight();
        int i = this.mGravity;
        if (i != 0) {
            if (i != 3) {
                if (i != 5) {
                    if (i != 51) {
                        if (i != 53) {
                            if (i == 80 || i == 83) {
                                if (this.mTriangle) {
                                    float f = (float) height;
                                    path.lineTo((float) width, f);
                                    path.lineTo(0.0f, f);
                                } else {
                                    float f2 = (float) height;
                                    path.lineTo((float) width, f2);
                                    path.lineTo((float) (width - this.mTextHeight), f2);
                                    path.lineTo(0.0f, (float) this.mTextHeight);
                                }
                                path.close();
                                canvas.drawPath(path, this.mBackgroundPaint);
                                canvas.save();
                            } else if (i == 85) {
                                if (this.mTriangle) {
                                    float f3 = (float) height;
                                    path.moveTo(0.0f, f3);
                                    float f4 = (float) width;
                                    path.lineTo(f4, f3);
                                    path.lineTo(f4, 0.0f);
                                } else {
                                    float f5 = (float) height;
                                    path.moveTo(0.0f, f5);
                                    path.lineTo(((float) this.mTextHeight) * 1.0f, f5);
                                    float f6 = (float) width;
                                    path.lineTo(f6, (float) this.mTextHeight);
                                    path.lineTo(f6, 0.0f);
                                }
                                path.close();
                                canvas.drawPath(path, this.mBackgroundPaint);
                                canvas.save();
                            } else {
                                throw new IllegalArgumentException("are you ok?");
                            }
                        }
                    }
                }
            }
            if (this.mTriangle) {
                path.lineTo(0.0f, (float) height);
                path.lineTo((float) width, 0.0f);
            } else {
                path.moveTo((float) width, 0.0f);
                path.lineTo(0.0f, (float) height);
                path.lineTo(0.0f, (float) (height - this.mTextHeight));
                path.lineTo((float) (width - this.mTextHeight), 0.0f);
            }
            path.close();
            canvas.drawPath(path, this.mBackgroundPaint);
            canvas.save();
        }
        if (this.mTriangle) {
            float f7 = (float) width;
            path.lineTo(f7, 0.0f);
            path.lineTo(f7, (float) height);
        } else {
            float f8 = (float) width;
            path.lineTo(f8, (float) height);
            path.lineTo(f8, (float) (height - this.mTextHeight));
            path.lineTo(((float) this.mTextHeight) * 1.0f, 0.0f);
        }
        path.close();
        canvas.drawPath(path, this.mBackgroundPaint);
        canvas.save();
    }

    private void drawText(Canvas canvas) {
        float f;
        float f2;
        float f3;
        float f4;
        float f5;
        int width = canvas.getWidth() - (this.mTextHeight / 2);
        int height = canvas.getHeight();
        int i = this.mTextHeight;
        int i2 = height - (i / 2);
        int i3 = i / 2;
        int i4 = this.mGravity;
        float f6 = -45.0f;
        if (i4 != 0) {
            if (i4 != 3) {
                if (i4 != 5) {
                    if (i4 != 51) {
                        if (i4 != 53) {
                            if (i4 == 80 || i4 == 83) {
                                Rect rect = new Rect(0, i3, width, i2 + i3);
                                RectF rectF = new RectF(rect);
                                TextPaint textPaint = this.mTextPaint;
                                String str = this.mText;
                                rectF.right = textPaint.measureText(str, 0, str.length());
                                rectF.bottom = this.mTextPaint.descent() - this.mTextPaint.ascent();
                                rectF.left += (((float) rect.width()) - rectF.right) / 2.0f;
                                rectF.top += (((float) rect.height()) - rectF.bottom) / 2.0f;
                                f2 = rectF.left;
                                f5 = rectF.top - this.mTextPaint.ascent();
                                f4 = ((float) width) / 2.0f;
                                f3 = (((float) i2) / 2.0f) + ((float) i3);
                                f = f5;
                                f6 = 45.0f;
                                canvas.rotate(f6, f4, f3);
                                canvas.drawText(this.mText, f2, f, this.mTextPaint);
                            } else if (i4 == 85) {
                                Rect rect2 = new Rect(i3, i3, width + i3, i2 + i3);
                                RectF rectF2 = new RectF(rect2);
                                TextPaint textPaint2 = this.mTextPaint;
                                String str2 = this.mText;
                                rectF2.right = textPaint2.measureText(str2, 0, str2.length());
                                rectF2.bottom = this.mTextPaint.descent() - this.mTextPaint.ascent();
                                rectF2.left += (((float) rect2.width()) - rectF2.right) / 2.0f;
                                rectF2.top += (((float) rect2.height()) - rectF2.bottom) / 2.0f;
                                f2 = rectF2.left;
                                f = rectF2.top - this.mTextPaint.ascent();
                                float f7 = (float) i3;
                                f4 = (((float) width) / 2.0f) + f7;
                                f3 = (((float) i2) / 2.0f) + f7;
                                canvas.rotate(f6, f4, f3);
                                canvas.drawText(this.mText, f2, f, this.mTextPaint);
                            } else {
                                throw new IllegalArgumentException("are you ok?");
                            }
                        }
                    }
                }
            }
            Rect rect3 = new Rect(0, 0, width, i2);
            RectF rectF3 = new RectF(rect3);
            TextPaint textPaint3 = this.mTextPaint;
            String str3 = this.mText;
            rectF3.right = textPaint3.measureText(str3, 0, str3.length());
            rectF3.bottom = this.mTextPaint.descent() - this.mTextPaint.ascent();
            rectF3.left += (((float) rect3.width()) - rectF3.right) / 2.0f;
            rectF3.top += (((float) rect3.height()) - rectF3.bottom) / 2.0f;
            float f8 = rectF3.left;
            f = rectF3.top - this.mTextPaint.ascent();
            f4 = ((float) width) / 2.0f;
            f3 = ((float) i2) / 2.0f;
            f2 = f8;
            canvas.rotate(f6, f4, f3);
            canvas.drawText(this.mText, f2, f, this.mTextPaint);
        }
        Rect rect4 = new Rect(i3, 0, width + i3, i2);
        RectF rectF4 = new RectF(rect4);
        TextPaint textPaint4 = this.mTextPaint;
        String str4 = this.mText;
        rectF4.right = textPaint4.measureText(str4, 0, str4.length());
        rectF4.bottom = this.mTextPaint.descent() - this.mTextPaint.ascent();
        rectF4.left += (((float) rect4.width()) - rectF4.right) / 2.0f;
        rectF4.top += (((float) rect4.height()) - rectF4.bottom) / 2.0f;
        f2 = rectF4.left;
        f5 = rectF4.top - this.mTextPaint.ascent();
        f4 = (((float) width) / 2.0f) + ((float) i3);
        f3 = ((float) i2) / 2.0f;
        f = f5;
        f6 = 45.0f;
        canvas.rotate(f6, f4, f3);
        canvas.drawText(this.mText, f2, f, this.mTextPaint);
    }

    public String getText() {
        return this.mText;
    }

    public void setText(int i) {
        setText(getResources().getString(i));
    }

    public void setText(String str) {
        if (!TextUtils.isEmpty(str) && !getText().equals(str)) {
            this.mText = str;
            invalidate();
        }
    }

    public int getTextColor() {
        return this.mTextPaint.getColor();
    }

    public void setTextColor(int i) {
        if (getTextColor() != i) {
            this.mTextPaint.setColor(i);
            invalidate();
        }
    }

    public float getTextSize() {
        return this.mTextPaint.getTextSize();
    }

    public void setTextSize(float f) {
        setTextSize(2, f);
    }

    public void setTextSize(int i, float f) {
        float applyDimension = TypedValue.applyDimension(i, f, getResources().getDisplayMetrics());
        if (getTextSize() != applyDimension) {
            this.mTextPaint.setTextSize(applyDimension);
            invalidate();
        }
    }

    public Typeface getTextStyle() {
        return this.mTextPaint.getTypeface();
    }

    public void setTextStyle(Typeface typeface) {
        if (getTextStyle() != typeface) {
            this.mTextPaint.setTypeface(typeface);
            invalidate();
        }
    }

    public int getColorBackground() {
        return this.mColorBackground;
    }

    public void setColorBackground(int i) {
        if (getColorBackground() != i) {
            this.mColorBackground = i;
            this.mBackgroundPaint.setColor(i);
            invalidate();
        }
    }

    public int getGravity() {
        return this.mGravity;
    }

    public void setGravity(int i) {
        if (this.mGravity != i) {
            this.mGravity = Gravity.getAbsoluteGravity(i, getResources().getConfiguration().getLayoutDirection());
            invalidate();
        }
    }

    public boolean isTriangle() {
        return this.mTriangle;
    }

    public void setTriangle(boolean z) {
        if (isTriangle() != z) {
            this.mTriangle = z;
            invalidate();
        }
    }

    private int getAccentColor() {
        TypedValue typedValue = new TypedValue();
        getContext().getTheme().resolveAttribute(R.attr.colorAccent, typedValue, true);
        return typedValue.data;
    }
}

File: decompiled/sources/com/hjq/widget/view/SlantedTextView.kt
package com.hjq.widget.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Rect
import android.graphics.RectF
import android.graphics.Typeface
import android.text.TextPaint
import android.text.TextUtils
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import androidx.core.view.GravityCompat
import com.hjq.widget.R

class SlantedTextView : View {
    private val mBackgroundPaint: Paint
    private var mColorBackground = 0
    private var mGravity = 0
    private var mText: String = ""
    private val mTextBounds = Rect()
    private var mTextHeight = 0
    private val mTextPaint: TextPaint
    private var mTriangle = false

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        val paint = Paint()
        mBackgroundPaint = paint
        paint.style = Paint.Style.FILL
        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_OVER)
        paint.isAntiAlias = true

        val textPaint = TextPaint(1)
        mTextPaint = textPaint
        textPaint.isAntiAlias = true

        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.SlantedTextView)
        setText(typedArray.getString(R.styleable.SlantedTextView_android_text))
        setTextSize(
            0,
            typedArray.getDimensionPixelSize(
                R.styleable.SlantedTextView_android_textSize,
                resources.getDimension(R.dimen.sp_12).toInt()
            ).toFloat()
        )
        setTextColor(typedArray.getColor(R.styleable.SlantedTextView_android_textColor, -0x1))
        setTextStyle(
            Typeface.defaultFromStyle(
                typedArray.getInt(R.styleable.SlantedTextView_android_textStyle, 0)
            )
        )
        setGravity(typedArray.getInt(R.styleable.SlantedTextView_android_gravity, GravityCompat.END))
        setColorBackground(
            typedArray.getColor(
                R.styleable.SlantedTextView_android_colorBackground,
                getAccentColor()
            )
        )
        setTriangle(typedArray.getBoolean(R.styleable.SlantedTextView_triangle, false))
        typedArray.recycle()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        mTextPaint.getTextBounds(mText, 0, mText.length, mTextBounds)
        mTextHeight = mTextBounds.height() + paddingTop + paddingBottom

        val widthMode = MeasureSpec.getMode(widthMeasureSpec)
        val measuredWidth = when (widthMode) {
            MeasureSpec.AT_MOST, MeasureSpec.UNSPECIFIED -> mTextBounds.width() + paddingLeft + paddingRight
            MeasureSpec.EXACTLY -> MeasureSpec.getSize(widthMeasureSpec)
            else -> 0
        }

        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
        val measuredHeight = when (heightMode) {
            MeasureSpec.AT_MOST, MeasureSpec.UNSPECIFIED -> mTextBounds.height() + paddingTop + paddingBottom
            MeasureSpec.EXACTLY -> MeasureSpec.getSize(heightMeasureSpec)
            else -> 0
        }

        val size = maxOf(measuredWidth, measuredHeight)
        setMeasuredDimension(size, size)
    }

    override fun onDraw(canvas: Canvas) {
        drawBackground(canvas)
        drawText(canvas)
    }

    private fun drawBackground(canvas: Canvas) {
        val path = Path()
        val width = canvas.width
        val height = canvas.height
        val gravity = mGravity
        if (gravity != 0) {
            if (gravity != 3) {
                if (gravity != 5) {
                    if (gravity != 51) {
                        if (gravity != 53) {
                            if (gravity == 80 || gravity == 83) {
                                if (mTriangle) {
                                    val h = height.toFloat()
                                    path.lineTo(width.toFloat(), h)
                                    path.lineTo(0.0f, h)
                                } else {
                                    val h = height.toFloat()
                                    path.lineTo(width.toFloat(), h)
                                    path.lineTo((width - mTextHeight).toFloat(), h)
                                    path.lineTo(0.0f, mTextHeight.toFloat())
                                }
                                path.close()
                                canvas.drawPath(path, mBackgroundPaint)
                                canvas.save()
                            } else if (gravity == 85) {
                                if (mTriangle) {
                                    val h = height.toFloat()
                                    path.moveTo(0.0f, h)
                                    val w = width.toFloat()
                                    path.lineTo(w, h)
                                    path.lineTo(w, 0.0f)
                                } else {
                                    val h = height.toFloat()
                                    path.moveTo(0.0f, h)
                                    path.lineTo((mTextHeight.toFloat()) * 1.0f, h)
                                    val w = width.toFloat()
                                    path.lineTo(w, mTextHeight.toFloat())
                                    path.lineTo(w, 0.0f)
                                }
                                path.close()
                                canvas.drawPath(path, mBackgroundPaint)
                                canvas.save()
                            } else {
                                throw IllegalArgumentException("are you ok?")
                            }
                        }
                    }
                }
            }
            if (mTriangle) {
                path.lineTo(0.0f, height.toFloat())
                path.lineTo(width.toFloat(), 0.0f)
            } else {
                path.moveTo(width.toFloat(), 0.0f)
                path.lineTo(0.0f, height.toFloat())
                path.lineTo(0.0f, (height - mTextHeight).toFloat())
                path.lineTo((width - mTextHeight).toFloat(), 0.0f)
            }
            path.close()
            canvas.drawPath(path, mBackgroundPaint)
            canvas.save()
        }

        if (mTriangle) {
            val w = width.toFloat()
            path.lineTo(w, 0.0f)
            path.lineTo(w, height.toFloat())
        } else {
            val w = width.toFloat()
            path.lineTo(w, height.toFloat())
            path.lineTo(w, (height - mTextHeight).toFloat())
            path.lineTo((mTextHeight.toFloat()) * 1.0f, 0.0f)
        }
        path.close()
        canvas.drawPath(path, mBackgroundPaint)
        canvas.save()
    }

    private fun drawText(canvas: Canvas) {
        var textX: Float
        var textY: Float
        var pivotX: Float
        var pivotY: Float

        val width = canvas.width - (mTextHeight / 2)
        val height = canvas.height
        val textHeight = mTextHeight
        val usableBottom = height - (textHeight / 2)
        val halfTextHeight = textHeight / 2
        val gravity = mGravity
        var angle = -45.0f

        if (gravity != 0) {
            if (gravity != 3) {
                if (gravity != 5) {
                    if (gravity != 51) {
                        if (gravity != 53) {
                            if (gravity == 80 || gravity == 83) {
                                val rect = Rect(0, halfTextHeight, width, usableBottom + halfTextHeight)
                                val rectF = RectF(rect)
                                rectF.right = mTextPaint.measureText(mText, 0, mText.length)
                                rectF.bottom = mTextPaint.descent() - mTextPaint.ascent()
                                rectF.left += (rect.width().toFloat() - rectF.right) / 2.0f
                                rectF.top += (rect.height().toFloat() - rectF.bottom) / 2.0f
                                textX = rectF.left
                                textY = rectF.top - mTextPaint.ascent()
                                pivotX = width.toFloat() / 2.0f
                                pivotY = (usableBottom.toFloat() / 2.0f) + halfTextHeight.toFloat()
                                angle = 45.0f
                                canvas.rotate(angle, pivotX, pivotY)
                                canvas.drawText(mText, textX, textY, mTextPaint)
                                return
                            } else if (gravity == 85) {
                                val rect = Rect(
                                    halfTextHeight,
                                    halfTextHeight,
                                    width + halfTextHeight,
                                    usableBottom + halfTextHeight
                                )
                                val rectF = RectF(rect)
                                rectF.right = mTextPaint.measureText(mText, 0, mText.length)
                                rectF.bottom = mTextPaint.descent() - mTextPaint.ascent()
                                rectF.left += (rect.width().toFloat() - rectF.right) / 2.0f
                                rectF.top += (rect.height().toFloat() - rectF.bottom) / 2.0f
                                textX = rectF.left
                                textY = rectF.top - mTextPaint.ascent()
                                val half = halfTextHeight.toFloat()
                                pivotX = (width.toFloat() / 2.0f) + half
                                pivotY = (usableBottom.toFloat() / 2.0f) + half
                                canvas.rotate(angle, pivotX, pivotY)
                                canvas.drawText(mText, textX, textY, mTextPaint)
                                return
                            } else {
                                throw IllegalArgumentException("are you ok?")
                            }
                        }
                    }
                }
            }

            val rect = Rect(0, 0, width, usableBottom)
            val rectF = RectF(rect)
            rectF.right = mTextPaint.measureText(mText, 0, mText.length)
            rectF.bottom = mTextPaint.descent() - mTextPaint.ascent()
            rectF.left += (rect.width().toFloat() - rectF.right) / 2.0f
            rectF.top += (rect.height().toFloat() - rectF.bottom) / 2.0f
            textX = rectF.left
            textY = rectF.top - mTextPaint.ascent()
            pivotX = width.toFloat() / 2.0f
            pivotY = usableBottom.toFloat() / 2.0f
            canvas.rotate(angle, pivotX, pivotY)
            canvas.drawText(mText, textX, textY, mTextPaint)
            return
        }

        val rect = Rect(halfTextHeight, 0, width + halfTextHeight, usableBottom)
        val rectF = RectF(rect)
        rectF.right = mTextPaint.measureText(mText, 0, mText.length)
        rectF.bottom = mTextPaint.descent() - mTextPaint.ascent()
        rectF.left += (rect.width().toFloat() - rectF.right) / 2.0f
        rectF.top += (rect.height().toFloat() - rectF.bottom) / 2.0f
        textX = rectF.left
        textY = rectF.top - mTextPaint.ascent()
        pivotX = (width.toFloat() / 2.0f) + halfTextHeight.toFloat()
        pivotY = usableBottom.toFloat() / 2.0f
        angle = 45.0f
        canvas.rotate(angle, pivotX, pivotY)
        canvas.drawText(mText, textX, textY, mTextPaint)
    }

    fun getText(): String {
        return mText
    }

    fun setText(resId: Int) {
        setText(resources.getString(resId))
    }

    fun setText(text: String?) {
        if (!TextUtils.isEmpty(text) && !getText().equals(text)) {
            mText = text!!
            invalidate()
        }
    }

    fun getTextColor(): Int {
        return mTextPaint.color
    }

    fun setTextColor(color: Int) {
        if (getTextColor() != color) {
            mTextPaint.color = color
            invalidate()
        }
    }

    fun getTextSize(): Float {
        return mTextPaint.textSize
    }

    fun setTextSize(size: Float) {
        setTextSize(2, size)
    }

    fun setTextSize(unit: Int, size: Float) {
        val applied = TypedValue.applyDimension(unit, size, resources.displayMetrics)
        if (getTextSize() != applied) {
            mTextPaint.textSize = applied
            invalidate()
        }
    }

    fun getTextStyle(): Typeface {
        return mTextPaint.typeface
    }

    fun setTextStyle(typeface: Typeface?) {
        if (getTextStyle() != typeface) {
            mTextPaint.typeface = typeface
            invalidate()
        }
    }

    fun getColorBackground(): Int {
        return mColorBackground
    }

    fun setColorBackground(color: Int) {
        if (getColorBackground() != color) {
            mColorBackground = color
            mBackgroundPaint.color = color
            invalidate()
        }
    }

    fun getGravity(): Int {
        return mGravity
    }

    fun setGravity(gravity: Int) {
        if (mGravity != gravity) {
            mGravity = Gravity.getAbsoluteGravity(gravity, resources.configuration.layoutDirection)
            invalidate()
        }
    }

    fun isTriangle(): Boolean {
        return mTriangle
    }

    fun setTriangle(triangle: Boolean) {
        if (isTriangle() != triangle) {
            mTriangle = triangle
            invalidate()
        }
    }

    private fun getAccentColor(): Int {
        val typedValue = TypedValue()
        context.theme.resolveAttribute(R.attr.colorAccent, typedValue, true)
        return typedValue.data
    }

    companion object {
        const val ROTATE_ANGLE: Int = 45
    }
}
